extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/KafkaRecordSupplier.java

```diff
@@ -0,0 +1,214 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class KafkaRecordSupplier implements RecordSupplier<Integer, Long>
+{
+  private final KafkaConsumer<byte[], byte[]> consumer;
+  private final Map<String, Object> consumerProperties;
+  private final ObjectMapper sortingMapper;
+  private boolean closed;
+
+  public KafkaRecordSupplier(
+      Map<String, Object> consumerProperties,
+      ObjectMapper sortingMapper
+  )
+  {
+    this.consumerProperties = consumerProperties;
+    this.sortingMapper = sortingMapper;
+    this.consumer = getKafkaConsumer();
+  }
+
+  @Override
+  public void assign(Set<StreamPartition<Integer>> streamPartitions)
+  {
+    consumer.assign(streamPartitions
+                        .stream()
+                        .map(x -> new TopicPartition(x.getStream(), x.getPartitionId()))
+                        .collect(Collectors.toSet()));
+    seekToEarliest(streamPartitions);
+  }
+
+  @Override
+  public void seek(StreamPartition<Integer> partition, Long sequenceNumber)
+  {
+    consumer.seek(new TopicPartition(partition.getStream(), partition.getPartitionId()), sequenceNumber);
+  }
+
+  @Override
+  public void seekToEarliest(Set<StreamPartition<Integer>> partitions)
+  {
+    consumer.seekToBeginning(partitions
+                                 .stream()
+                                 .map(e -> new TopicPartition(e.getStream(), e.getPartitionId()))
+                                 .collect(Collectors.toList()));
+  }
+
+  @Override
+  public void seekToLatest(Set<StreamPartition<Integer>> partitions)
+  {
+    consumer.seekToEnd(partitions
+                           .stream()
+                           .map(e -> new TopicPartition(e.getStream(), e.getPartitionId()))
+                           .collect(Collectors.toList()));
+  }
+
+  @Override
+  public Set<StreamPartition<Integer>> getAssignment()
+  {
+    Set<TopicPartition> topicPartitions = consumer.assignment();
+    return topicPartitions
+        .stream()
+        .map(e -> new StreamPartition<>(e.topic(), e.partition()))
+        .collect(Collectors.toSet());
+  }
+
+  @Nonnull
+  @Override
+  public List<OrderedPartitionableRecord<Integer, Long>> poll(long timeout)
+  {
+    List<OrderedPartitionableRecord<Integer, Long>> polledRecords = new ArrayList<>();
+    for (ConsumerRecord<byte[], byte[]> record : consumer.poll(timeout)) {
+      polledRecords.add(new OrderedPartitionableRecord<>(
+          record.topic(),
+          record.partition(),
+          record.offset(),
+          record.value() == null ? null : ImmutableList.of(record.value())
+      ));
+    }
+    return polledRecords;
+  }
+
+  @Override
+  public Long getLatestSequenceNumber(StreamPartition<Integer> partition)
+  {
+    Long currPos = consumer.position(new TopicPartition(partition.getStream(), partition.getPartitionId()));
```

Comment:
why not bump the kafka version to 0.11 and use https://kafka.apache.org/0110/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#beginningOffsets(java.util.Collection) and https://kafka.apache.org/0110/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#endOffsets(java.util.Collection)

Reply:
There's another PR in the works -- #6496 -- to upgrade to Kafka 2.x, from @surekhasaharan, that we were planning to do this in after this current PR has been merged (to avoid conflicts). There were other changes needed too in order to make the upgrade work smoothly (check out that PR). So IMO it makes sense to leave this as is, & we can make the new Kafka API changes over in #6496.